qiita_db/processing_job.py

```diff
@@ -6,21 +6,295 @@
+        # stop and the Watcher process will exit.
+        self.event.set()
+        # Here, it is assumed that we are running this from the main
+        # context. By joining(), we're waiting for the Watcher process to
+        # end before returning from this method.
+        self.join()
+
+
+def launch_local(env_script, start_script, url, job_id, job_dir):
+
+    # launch_local() differs from launch_torque(), as no Watcher() is used.
+    # each launch_local() process will execute the cmd as a child process,
+    # wait, and update the database once cmd has completed.
+    #
+    # As processes are lighter weight than jobs, this should be fine.
+    # This is how the current job model works locally.
+
+    cmd = [start_script, url, job_id, job_dir]
+
+    # When Popen() executes, the shell is not in interactive mode,
+    # so it is not sourcing any of the bash configuration files
+    # We need to source it so the env_script are available
+    cmd = "bash -c '%s; %s'" % (env_script, ' '.join(cmd))
+
+    # Popen() may also need universal_newlines=True
+    proc = Popen(cmd, shell=True, stdout=PIPE, stderr=PIPE)
+
+    # Communicate pulls all stdout/stderr from the PIPEs
+    # This call waits until cmd is done
+    stdout, stderr = proc.communicate()
+
+    # proc.returncode will be equal to None if the process hasn't finished
+    # yet. If cmd was terminated by a SIGNAL, it will be a negative value.
+    # (*nix platforms only)
+    error = None
+
+    if proc.returncode != 0:
+        error = "error from launch_local when launching cmd='%s'" % cmd
+        error = "%s\n%s\n%s" % (error, stdout, stderr)
+
+        # Forcing the creation of a new connection
+        qdb.sql_connection.create_new_transaction()
+        ProcessingJob(job_id).complete(False, error=error)
+
+
+def launch_torque(env_script, start_script, url, job_id, job_dir,
+                  dependent_job_id, resource_params):
+
+    # note that job_id is Qiita's UUID, not a Torque job ID
+    cmd = [start_script, url, job_id, job_dir]
+
+    # generating file contents to be used with qsub
+    lines = []
+
+    # TODO: is PBS_JOBID is being set correctly?
+    lines.append("echo $PBS_JOBID")
+
+    # TODO: revisit below
+    lines.append("source ~/.bash_profile")
+    lines.append(env_script)
+    lines.append(' '.join(cmd))
+
+    # writing the file to be used with qsub
+    create_nested_path(job_dir)
+
+    fp = join(job_dir, '%s.txt' % job_id)
+    open(fp, 'w').write("\n".join(lines))
+
+    qsub_cmd = []
+
+    if dependent_job_id:
+        # note that a dependent job should be submitted before the 'parent' job
+        # ends, most likely. Torque doesn't keep job state around forever, and
+        # creating a dependency on a job already completed has not been tested.
+        qsub_cmd.append("qsub -W depend=afterok:%s" % dependent_job_id)
+    else:
+        qsub_cmd.append("qsub")
+
+    qsub_cmd.append("%s %s" % (resource_params, fp))
```

Comment:
I would suggest separating the flags and the values, i.e. ['%s', '%s'] and using .extend in the qsub_cmd list. This will escape any problematic characters in the directory names.

Reply:
good call - will do!